File: src/CSharp/46.Permutations.cs
using System;
using System.Text;
using System.Collections.Generic;


namespace PeefyLeetCode.Permutations
{
public class Solution {
    void dfs(IList<IList<int>> ans, int[] visited, IList<int> path, int[] nums){
        if (path.Count == nums.Length)
            ans.Add(new List<int>(path));
        else{
            for (int i = 0;i < nums.Length;++i){
                if (visited[i] == 0){
                    visited[i] = 1;
                    path.Add(nums[i]);
                    dfs(ans, visited, path, nums);
                    path.RemoveAt(path.Count - 1);
                    visited[i] = 0;
                }
            }
        }
    }
    public IList<IList<int>> Permute(int[] nums) {
        IList<IList<int>> ans = new List<IList<int>>();
        int[] visited = new int[nums.Length];      
        IList<int> path = new List<int>();
        dfs(ans, visited, path, nums);
        return ans;
    }
}
}

File: src/Python/46.Permutations.py
from copy import deepcopy

class Solution:
    def permute(self, nums: 'List[int]') -> 'List[List[int]]':
        ans = []
        tmp = []
        m = {}
        for num in nums:
            m[num] = 0
        def recursive(i, nums, ans, tmp, m):
            if len(tmp) == len(nums):
                return ans.append(deepcopy(tmp))
            for j in range(len(nums)):
                if m[nums[j]] == 0:
                    tmp.append(nums[j])
                    m[nums[j]] += 1
                    recursive(j + 1, nums, ans, tmp, m)
                    tmp.pop()
                    m[nums[j]] -= 1
        recursive(0, nums, ans, tmp, m)
        return ans

if __name__ == "__main__":
    solution = Solution()
    print(solution.permute([1, 2, 3]))
    print(solution.permute([1]))

   